### DISCUSSION TITLE
Image Carousel?

### DISCUSSION THREAD

**Discussion by MrGonzo1024** [Budibase/budibase#4985] [Help]
Hi all,

Loving Budibase! 

Has anyone have success on implementing an image carousel?  I didn't see any discussions on it.

Thanks,
Gil.

**Comment by melohagan** [comment]
Hi @MrGonzo1024 

There is no image carousel component, however you can build a basic component in Budibase:

![image_carousel_demo](https://user-images.githubusercontent.com/101575380/160257261-58db77f2-e415-4757-b515-a5f1017363ad.gif)

Doing this requires just a little JavaScript and use of the <a href="https://docs.budibase.com/docs/app-state">App State</a>

### Step 1 - Data

Add a table with indexed image URL paths:
<img width="1172" alt="Screenshot 2022-03-26 at 21 13 30" src="https://user-images.githubusercontent.com/101575380/160257324-352376dd-c398-4897-af16-6d6e6f3360d0.png">

### Step 2 - Design

Create the following structure:
<img width="264" alt="Screenshot 2022-03-26 at 21 16 10" src="https://user-images.githubusercontent.com/101575380/160257421-b4a7b103-88e0-4f78-9388-f3b4d0f81bb2.png">

We want the table data to be provided for both the images and the button. 
The data provider has an ascending sort on the index column.
The Image Container should be 'Row' direction to get the images to be arranged horizontally across. 
The centre image is also sized to be larger than the flanks.  

### Step 3 - Actions

Define an action for the **Next** button:
<img width="928" alt="Screenshot 2022-03-26 at 21 27 36" src="https://user-images.githubusercontent.com/101575380/160257742-a2f666c2-3491-4c7e-a182-f5cdc0016db7.png">

Binding Value:
```js
let imageIndex = $("State.ImageIndex");

if (!imageIndex) {
	imageIndex = 1;
}
else if (imageIndex === $("Data Provider.Rows Length")-1) {
	imageIndex = 0;
}
else {
	imageIndex++;
}

return imageIndex; 
```
>This app state will keep track of the current active image shown, and be incremented when the button is clicked.

----
Then add JavaScript bindings to the Image URLs as follows:

**Image Left**
```js
let imageIndex = $("State.ImageIndex");
let rows = $("Data Provider.Rows");

if (!imageIndex) {
	return rows.filter(row => row.Index === $("Data Provider.Rows Length")-1)[0].URL;
}

return rows.filter(row => row.Index === imageIndex-1)[0].URL;
```
>Show the previous image (or loop to the last if the first image is active)

**Image Center**
```js
let imageIndex = $("State.ImageIndex");
let rows = $("Data Provider.Rows");

if (!imageIndex) {
	return rows.filter(row => row.Index === 0)[0].URL;
}

return rows.filter(row => row.Index === imageIndex)[0].URL;
```
>Show the active current image (or the first if the next button has not been clicked yet)

**Image Right**
```js
let imageIndex = $("State.ImageIndex");
let rows = $("Data Provider.Rows");

if (!imageIndex) {
	return rows.filter(row => row.Index === 1)[0].URL;
}
else if (imageIndex === $("Data Provider.Rows Length")-1) {
	return rows.filter(row => row.Index === 0)[0].URL;
}

return rows.filter(row => row.Index === imageIndex+1)[0].URL;
```
>Show the next image (or loop to the first if the last image is active)
----

To explore further, here is the exported app:
[image_carousel-export-1648328704823.txt](https://github.com/Budibase/budibase/files/8356413/image_carousel-export-1648328704823.txt)

**Reply by MrGonzo1024** [reply]
Thank you so much @melohagan ! I will try it.
